### DISCUSSION TITLE
Rigth way to use linq2db

### DISCUSSION THREAD

**Discussion by ImoutoChan** [linq2db/linq2db#3397] [Q&A] [answered]
Hello, I'm a little confused about the right way to use linq2db.

I have a general web API application on .NET Core 5.0. My code has a context class inherited from `DataConnection`. I register it with `AddLinqToDbContext `method. I inject my context class into the constructor and use it this way. I don't have singletons, all my services are transient or (rarely) scoped. Everything is called from scoped context either by asp net core or quartz.

So, my questions are:
1. In my context class I have to use this construction `(this as IDataContext).CloseAfterUse = true;` because without it transactions don't work. Rollback completes successfully, but part of the data created in this transaction is still in the database. But with `CloseAfterUse` everything works fine. Is it a bug or I don't understand something? I create transactions using TransactionScope.

2. My application under heavy load begins to throw exceptions about timeout on trying to get a connection from the connection pool. I think `CloseAfterUse` exhausts the connection pool earlier. Is it rational to increase the size of my connection pool? Or does it mean I have a connection leak somewhere in my code?

P.s. I use `linq2db` and `linq2db.AspNet` packages with version `4.0.0-preview.3`

**Comment by MaceWindu** [comment, marked as answer]
1. If you use `DataConnection` per-request, you should use `CloseAfterUse = true` to release connection back to pool between transactions/non-transactional queries to reduce(!) chances to exhause connection pool due to connection being held while not used. Still it doesn't protect you from pool exhaustion and you should configure pool size to match your expected/manageable load. It doesn't sound like you have connection leak as it manifests only on heavy load.
2. Regarding transaction issues with TransactionScope I don't really know what issue you face. Maybe sample code will help to understand your issue.
